server/selection.go

```diff
@@ -9,6 +18,161 @@ type BroadcastSessionsSelector interface {
 	Clear()
 }
 
+type sessHeap []*BroadcastSession
+
+func (h sessHeap) Len() int {
+	return len(h)
+}
+
+func (h sessHeap) Less(i, j int) bool {
+	return h[i].LatencyScore < h[j].LatencyScore
+}
+
+func (h sessHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *sessHeap) Push(x interface{}) {
+	sess := x.(*BroadcastSession)
+	*h = append(*h, sess)
+}
+
+func (h *sessHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	sess := old[n-1]
+	old[n-1] = nil
+	*h = old[:n-1]
+
+	return sess
+}
+
+func (h *sessHeap) Peek() interface{} {
+	if h.Len() == 0 {
+		return nil
+	}
+
+	return (*h)[0]
+}
+
+type stakeReader interface {
+	Stakes(addrs []ethcommon.Address) (map[ethcommon.Address]*big.Int, error)
+}
+
+// MinLSSelector selects the next BroadcastSession with the lowest latency score if it is good enough.
+// Otherwise, it selects a session that does not have a latency score yet
+type MinLSSelector struct {
+	unknownSessList []*BroadcastSession
+	knownSessList   *sessHeap
+
+	stakeRdr stakeReader
+
+	goodEnoughLS float64
+}
+
+// NewMinLSSelector returns an instance of MinLSSelector configured with a good enough latency score
+func NewMinLSSelector(stakeRdr stakeReader, goodEnoughLS float64) *MinLSSelector {
+	knownSessList := &sessHeap{}
+	heap.Init(knownSessList)
+
+	return &MinLSSelector{
+		knownSessList: knownSessList,
+		stakeRdr:      stakeRdr,
+		goodEnoughLS:  goodEnoughLS,
+	}
+}
+
+// Add adds the sessions to the selector's list of sessions without a latency score
+func (s *MinLSSelector) Add(sessions []*BroadcastSession) {
+	s.unknownSessList = append(s.unknownSessList, sessions...)
+}
+
+// Complete adds the session to the selector's list sessions with a latency score
+func (s *MinLSSelector) Complete(sess *BroadcastSession) {
+	heap.Push(s.knownSessList, sess)
+}
+
+// Select returns the session with the lowest latency score if it is good enough.
+// Otherwise, a session without a latency score yet is returned
+func (s *MinLSSelector) Select() *BroadcastSession {
+	if s.knownSessList.Len() == 0 {
+		return s.unknownSessListSelect()
+	}
+
+	// knownSessList is not empty at this point so we know minSess is not nil
+	minSess := s.knownSessList.Peek().(*BroadcastSession)
```

Comment:
> Not sure how much of a concern this is currently, but if there is a small chance that the list could become empty between the previous Len() and this Peek() (and even the Pop) then it might be safer to do something like the following

Removed the `Len()` call in favor of just calling `Peek()` and checking for a nil session in 711f9ea.

> Even though this sort of behavior probably isn't likely right now (are we sure the selector won't be used concurrently right now?)

The selector's caller (i.e. `BroadcastSessionsManager`) should implement concurrency safety. Added a comment in 711f9ea noting that `MinLSSelector` is not concurrency safe.

> Or the session could be popped unconditionally, and re-pushed in case the latency score is not good enough?

While I like this approach from a defensive programming standpoint, I don't like that it requires 1 heap pop + 1 heap push (`O(lgn)` cost for each) instead of just 1 heap pop whenever a known session's latency score is not good enough. Since we're offloading the responsibility of concurrency safety to the selector's caller, I'm inclined not to use this defensive programming approach since its slightly more expensive.

Reply:
> I don't like that it requires 1 heap pop + 1 heap push (O(lgn) cost for each) instead of just 1 heap pop whenever a known session's latency score is not good enough

Note that N is currently quite small - IIRC, with our current parameters, it maxes out around N=8 orchestrators [1].

For the case where all orchestrators are slow but otherwise performing well enough to establish a steady state, the heap is never pushed more than N times. If the network is performing well, then the additional heap push will never happen [2]. Amortized across all invocations, the cost of the additional heap push should be small.

While I do prefer the defensive approach, will leave it up to you.

[1] HTTPTimeout.Seconds() = 8, SegLen.Seconds() = 2   https://github.com/livepeer/go-livepeer/blob/7fcf97071963657c96176188b398fe9466ba0a6a/server/broadcast.go#L167-L168

[2] The current approach of only considering the most recent latency score all but ensures that well-performing Os will get penalized by transient issues. So right now, the latency selector's view of network health is going to be a bit sensitive for a while: not only will we get additional heap pushes, overall latency may get worse due to mostly-fast Os being mistakenly placed at the back of the queue where it's difficult to recover.

Reply:
I was originally going to go with the defensive approach, but I realized that the current implementation of `sessHeap` will panic when `heap.Pop()` and `len(sessHeap) = 0` due to an index out of range error triggered by calling `sessHeap.Swap()`.  We could update the implementations of `sessHeap.Swap()` and `sessHeap.Pop()` so that `heap.Pop()` returns nil when `len(sessHeap) = 0` but this would require changing `sessHeap.Swap()` to no-op when `len(sessHeap) = 0` which feels a bit weird to me. Furthermore, this change would introduce the length check into `sessHeap.Swap()` which means we still have the length check it is just executed at a different point in the code. If we're going to need the length check no matter what, I'd prefer to just have the caller of `heap.Pop()` do the length check instead of doing it in `sessHeap.Swap()` as this avoids adding a no-op in `sessHeap.Swap()` when `len(sessHeap) = 0`.

> The current approach of only considering the most recent latency score all but ensures that well-performing Os will get penalized by transient issues.

Good point. Perhaps we can address this with #1232

Reply:
Ah I see what you mean - it's because the golang implementation is stupid:

https://golang.org/src/container/heap/heap.go?s=2190:2223#L50

OK then.